field_utils/FieldDelta.F90

```diff
@@ -0,0 +1,493 @@
+         if (.not. present(new_num_levels)) return
+
+         call MAPL_InfoGetInternal(field, KEY_NUM_LEVELS, value=current_num_levels, _RC)
+         call MAPL_InfoGetInternal(field, KEY_VLOC, value=vloc, _RC)
+
+         ! Surface fields are not impacted by change in vertical grid
+         _RETURN_IF(vloc == 'VERTICAL_DIM_NONE')
+
+         new_array = new_array .or. (this%num_levels /= current_num_levels)
+
+         select case (vloc)
+         case ('VERTICAL_DIM_CENTER')
+            ungriddedUBound(1) = this%num_levels
+         case ('VERTICAL_DIM_EDGE')
+            ungriddedUBound(1) = this%num_levels + 1
+         case default
+            _FAIL('unsupported vertical location: '//vloc)
+         end select
+
+         _RETURN(_SUCCESS)
+      end subroutine select_ungriddedUbound
+      
+   end subroutine reallocate_field
+
+
+   subroutine reallocate_fields(this, fieldList, ignore, rc)
+      class(FieldDelta), intent(in) :: this
+      type(ESMF_Field), intent(inout) :: fieldList(:)
+      character(*), intent(in) :: ignore
+      integer, optional, intent(out) :: rc
+
+      integer :: status
+      integer :: i
+
+      do i = 1, size(fieldList)
+         call this%reallocate_field(fieldList(i), ignore, _RC)
+      end do
+
+      _RETURN(_SUCCESS)
+   end subroutine reallocate_fields
+
+   ! TODO - delete when next ESMF release provides support.
+   
+   impure elemental logical function ESMF_GeomEqual(geom1, geom2)
+      type(ESMF_Geom), intent(in) :: geom1, geom2
+
+      type(ESMF_GeomType_Flag) :: geomtype1, geomtype2
+      type(ESMF_Grid) :: grid1, grid2
+      type(ESMF_LocStream) :: locstream1, locstream2
+      type(ESMF_Mesh) :: mesh1, mesh2
+      type(ESMF_XGrid) :: xgrid1, xgrid2
+      
+      ESMF_GeomEqual = .false.
+
+      call ESMF_GeomGet(geom1, geomtype=geomtype1)
+      call ESMF_GeomGet(geom2, geomtype=geomtype2)
+
+      if (geomtype1 /= geomtype2) return
+      
+      if (geomtype1 == ESMF_GEOMTYPE_GRID) then
+         call ESMF_GeomGet(geom1, grid=grid1)
+         call ESMF_GeomGet(geom2, grid=grid2)
+         ESMF_GeomEqual = (grid1 == grid2)
+         return
+      end if
+
+      if (geomtype1 == ESMF_GEOMTYPE_LOCSTREAM) then
+         call ESMF_GeomGet(geom1, locstream=locstream1)
+         call ESMF_GeomGet(geom2, locstream=locstream2)
+         ESMF_GeomEqual = (locstream1 == locstream2)
+         return
+      end if
+
+      if (geomtype1 == ESMF_GEOMTYPE_MESH) then
+         call ESMF_GeomGet(geom1, mesh=mesh1)
+         call ESMF_GeomGet(geom2, mesh=mesh2)
+         ESMF_GeomEqual = (mesh1 == mesh2)
+         return
+      end if
+
+      if (geomtype1 == ESMF_GEOMTYPE_XGRID) then
+         call ESMF_GeomGet(geom1, xgrid=xgrid1)
+         call ESMF_GeomGet(geom2, xgrid=xgrid2)
+         ESMF_GeomEqual = (xgrid1 == xgrid2)
+         return
+      end if
+      
+   end function ESMF_GeomEqual
+
+
+   impure elemental logical function ESMF_GeomNotEqual(geom1, geom2)
```

Comment:
This seems pure since geom1 and geom2 are intent(in).

Reply:
The ESMF calls are not themselves PURE.  So ...